[XLA] Add half as supported type for TruncatedNormal op definition

DavidNorman opened the pull request:
Add the half type as a supported type for the truncated normal random op.  Backends can still decide on a per-backend basis whether they support that type or not.

DavidNorman committed:
Add half as supported type for TruncatedNormal op definition

sanjoy commented:
Do you mind adding a test case to tensorflow/compiler/tests/random_ops_test.py ?

DavidNorman commented:
sure thing

DavidNorman commented:
Hi,

actually it already does test half types:

```
  def _random_types(self):
    return set(self.numeric_types) - set(
        self.complex_types) - {np.uint64, np.int64, np.uint8, np.int8}
```

sanjoy commented:
> Hi,
> 
> actually it already does test half types:

If there was a test it would have been failing. :)

I think you only need to change this line (and remove the TODO): https://github.com/tensorflow/tensorflow/blob/b1d7145be23670dc412659f36d62996d024ef709/tensorflow/compiler/tests/random_ops_test.py#L124

DavidNorman commented:
you are absolutely right.  I had checked the tests like:

```
  def testRandomNormalVariance(self):
    for dtype in self._random_types() & self.float_types:

```

but the `testTruncatedNormalIsInRange` is different.  will patch.

DavidNorman commented:
obviously i've only tested on the CPU and IPU backends, it could be that allowing all float types in that test will cause the GPU or TPU to get upset.  hopefully not.

DavidNorman commented:
i see. the 'truncated normal is not constant' doesn't test all types either. you are quite right - there were no tests at all.

DavidNorman committed:
Enable f16 tests for truncated normal where appropriate

Summary: Add float16 to the truncated normal random test

Test Plan: all tests ok

Reviewers: georgep, jakeh, stephenm, romand

Reviewed By: jakeh

Differential Revision: https://phabricator.sourcevertex.net/D12416

DavidNorman committed:
Merge branch 'master' into add-half-to-truncated-normal-op-def

DavidNorman commented:
done :)

DavidNorman commented on tensorflow/compiler/tests/random_ops_test.py:
```diff
@@ -116,12 +116,22 @@ def testTruncatedNormalIsNotConstant(self):
-    count = 10000000
-    # TODO(b/34339814): make this test work with 16 bit float types.
-    for dtype in self._random_types() & {dtypes.float32, dtypes.float64}:
+    count = 1000000
```
note: numpy gives an error with 10,000,000 reduction during the tests at the end. this count seems to provide an adequate level of accuracy.

sanjoy replied:
Can you be more specific -- what's the error?  And can you please add this as a comment?

DavidNorman commented on tensorflow/compiler/tests/random_ops_test.py:
```diff
@@ -116,12 +116,22 @@ def testTruncatedNormalIsNotConstant(self):
+
+      # TODO: CPU backend doesn't produce a valid distribution. Remove this
+      # check when it does.
+      if (self.device in ["XLA_GPU", "XLA_CPU"
```
I added this because the CPU doesn't give completely valid values for half.  i'm guessing that this is the same issue as is referenced on line 103

DavidNorman replied:
note though that the CPU does pass the 'truncated normal random values are different' test which has been enabled.

sanjoy commented on tensorflow/compiler/tests/random_ops_test.py:
```diff
@@ -116,12 +116,22 @@ def testTruncatedNormalIsNotConstant(self):
+    # There is no double precision float kernel for truncated normal
+    for dtype in self._random_types() & self.float_types - {np.float64}:
+
+      # TODO: CPU backend doesn't produce a valid distribution. Remove this
```
The comment says CPU  but the code says both CPU and GPU?

If you're being conservative because you don't have a GPU to run the tests on, can you please add a `TODO(sanjoy)` to enable the test if it passes?

DavidNorman replied:
sure thing.  you are right about being conservative.  given the comments and code further up the file, it seems likely that it will not work on the GPU.

DavidNorman replied:
ok - i checked the count thing again and it is ok as it was.  not sure why i thought that numpy wouldn't handle it

sanjoy replied:
Can you be more specific about the errors you're seeing on the CPU backend?

 - Do they happen with both half and bfloat16?
 - What asserts fail?

DavidNorman committed:
Increase count again to provide correct delta

DavidNorman committed:
Merge branch 'master' into add-half-to-truncated-normal-op-def

DavidNorman committed:
Fix message post review

DavidNorman commented:
@gbaned  I think I have sorted all of the requested changes here.

sanjoy commented on tensorflow/compiler/tests/random_ops_test.py:
```diff
@@ -116,12 +116,24 @@ def testTruncatedNormalIsNotConstant(self):
       return random_ops.truncated_normal(shape=[2], dtype=dtype)
 
-    self._testRngIsNotConstant(rng, dtypes.float32)
+    # There is no double precision float kernel for truncated normal
```
End comments with period.

sanjoy commented on tensorflow/compiler/tests/random_ops_test.py:
```diff
@@ -116,12 +116,24 @@ def testTruncatedNormalIsNotConstant(self):
+
+      # TODO: CPU backend doesn't produce a valid distribution. Remove this
+      # check when it does.
+      # TODO(sanjay): remove the check for XLA_GPU if the tests work on the
```
Should be `TODO(sanjoy)`

DavidNorman replied:
sure thing.  will do

sanjoy commented on tensorflow/compiler/tests/random_ops_test.py:
```diff
@@ -116,12 +116,24 @@ def testTruncatedNormalIsNotConstant(self):
-    # TODO(b/34339814): make this test work with 16 bit float types.
-    for dtype in self._random_types() & {dtypes.float32, dtypes.float64}:
+
+    # There is no double precision float kernel for truncated normal
```
End comments with period.

DavidNorman replied:
will do

sanjoy reviewed (changes requested):
Only minor stuff.

DavidNorman committed:
Merge branch 'master' into add-half-to-truncated-normal-op-def

DavidNorman committed:
Add comment and check for backend incompatibility

DavidNorman commented:
Hi,

i had to resolve an issue with this test changing slightly.  it seems that someone has added double support, so that is now in, but i have left in the checks for CPU/GPU w.r.t. halves.  

i left the TODO(sanjoy) in, even though there is a separate TODO now for getting the test working with halves.

gbaned commented:
@DavidNorman Could you please check Ubuntu Sanity errors? Thanks!

DavidNorman committed:
Merge branch 'master' into add-half-to-truncated-normal-op-def

DavidNorman committed:
Fix lint issue

DavidNorman commented:
@gbaned done :)

DavidNorman committed:
Merge branch 'master' into add-half-to-truncated-normal-op-def

DavidNorman commented:
i've had to abandon the test changes.  there is someone inside google who is also working on getting half formats into the random test, and it keeps conflicting.

please consider merging this without the test changes.  

this is the code: b/34339814

DavidNorman commented:
I've realised I can achieve the same result as this by overriding the TF2XLA implementation of truncated normal and specifying a backend-specific implementation of truncated normal.

closing this as no longer necessary.
